dart/dynamics/ShapeFrame.h

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (c) 2016, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Jeongseok Lee <jslee02@gmail.com>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_DYNAMICS_SHAPEFRAME_H_
+#define DART_DYNAMICS_SHAPEFRAME_H_
+
+#include <Eigen/Dense>
+
+#include "dart/common/Signal.h"
+#include "dart/common/AddonWithVersion.h"
+#include "dart/common/SpecializedAddonManager.h"
+#include "dart/dynamics/FixedFrame.h"
+#include "dart/dynamics/TemplatedJacobianNode.h"
+#include "dart/dynamics/EllipsoidShape.h"
+
+namespace dart {
+namespace dynamics {
+
+class VisualAddon;
+class CollisionAddon;
+class DynamicsAddon;
+class ShapeFrame;
+
+namespace detail {
+
+struct VisualDataProperties
+{
+  /// Color for the primitive shape
+  Eigen::Vector4d mRGBA;
+
+  bool mUseDefaultColor;
+
+  /// True if this shape node should be kept from rendering
+  bool mHidden;
+
+  /// Constructor
+  VisualDataProperties(
+      const Eigen::Vector4d& color = Eigen::Vector4d(0.5, 0.5, 1.0, 1.0),
+      const bool hidden = false);
+
+  /// Destructor
+  virtual ~VisualDataProperties() = default;
+};
+
+struct CollisionDataProperties
```

Comment:
I wonder if we should be able to embed information about which bodies to ignore collisions against. For example, maybe there are two bodies in a Skeleton which you always want to ignore collisions between even though they don't technically qualify as adjacent. You could have a ``std::set<std::string>`` containing the names of those bodies in here. Using the body names (and not something like their pointers) allows this information to be trivially cloned and copied. Body names within a Skeleton are guaranteed to be unique, so there's no need to worry about ambiguity. Same thing if we want to specify other ``ShapeNode``s to ignore collisions against.

It gets much trickier if we want to ignore collisions against bodies in other Skeletons. We could have a ``std::map<std::string, std::set<std::string> >`` where the map key is the name of the Skeleton and the set of strings are names of the bodies within that Skeleton. But this would make it difficult to specify a pattern like "I want to ignore all bodies with name ``x`` in Skeletons whose names begin with ``y``".

I know @jslee02 is working on Collision Groups now. I wonder if we could have a ``std::string`` (or something similar) in here that tags the collision group name for this ``ShapeFrame``, and then we could have some way of defining the properties of the collision group elsewhere (in a way that's cloneable).